### PR TITLE
Reload Mods button changes and available mod updates indicator

### PR DESCRIPTION
**Pull request by ghost**
### What is the new feature?
Aims to address #1386

- "Reload Mods" button renamed to "Force Reload All"
- "Force Reload All" button only shows when in developer mode
- New text on the "Mod Browser" entry on main menu to indicate number of available mod updates

### Why should this be part of tModLoader?
-  Button changes were suggested in Discord - I believe this reload button shouldn't really be useful to non-dev users anymore?
- Update reminder is useful to prevent users forgetting to update mods. This will still be useful post-steam-workshop for GOG users.

### Are there alternative designs?
An icon could be used for indicating available mod updates - either some !-like indicator or perhaps a number in a circle. But I wasn't sure how to do this, and the text is clearer in meaning.

### Screenshots
- Mod Update Available
image
image

- Mods Menu (Developer Mode = True)
image

- Mods Menu (Developer Mode = False)
image

### CODE REVIEW

**Review by Solxanich — COMMENTED** [review]
Just a quick formatting review.

File: patches/tModLoader/Terraria/ModLoader/ModLoader.cs [hunk 1]
```diff
@@ -195,6 +202,86 @@ private static void Load(CancellationToken token = default)
...
+					using (var client = new WebClient()) {
+						ServicePointManager.ServerCertificateValidationCallback = (sender, certificate, chain, policyErrors) => { return true; };
+						client.UploadValuesCompleted += UpdateCheckComplete;
+						client.UploadValuesAsync(new Uri(url), "POST", values);
+					}
+				}, _cts.Token);
+			}
+			catch (Exception e) {
+				// Fail silently, it doesn't really matter if the update indicator is missing
+				Logging.tML.Debug("Error checking for mod updates.");
```
[the remark this answers is not in the document]
Comment (jackbondpreston) [reply patches/tModLoader/Terraria/ModLoader/ModLoader.cs:232]: Is localisation necessary for debug logs? I couldn't find any existing examples of these having been localisated.

[the remark this answers is not in the document]
Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/ModLoader/ModLoader.cs:232]: We're a bit inconsistent on these ones. I'd say it's not necessary outside of ModCompile

**Comment by direwolf420** [comment]
I'm not a friend of the main menu having a "YOU DON'T HAVE X AMOUNT OF MODS UPDATED YOU FOOL" in the middle of the screen, but that's just me.

**Comment by Steviegt6** [comment]
> I'm not a big fan of the main menu having a "YOU DON'T HAVE X AMOUNT OF MODS UPDATED YOU FOOL" in the middle of the screen, but that's just me.

+1
The original proposal's design was incredibly more elegant in concept. I don't feel like we need "(x Update(s) Available) in the middle of our screen acting as an eye sore just for the sake of it. Drawing a little symbol wouldn't be challenging (would be even easier if my Menu Buttons PR was merged, SMH), or you could just change the color or something and give the button some mouse text on hover. Not sure of what would look the best, but the current design stands out in a bad way.

**Comment by jackbondpreston** [comment]
I'm happy to give the original proposal a shot if anyone can give guidance on where/how to go about drawing an icon overlaying the main menu.

Re: locking reload mods, it was discussed on discord and suggested to do this - I'm not totally in the know about this but I think it seemed like mods are automatically reloaded like you mentioned? Someone else can probably shine light on this, I'm happy to revert it if not.

**Comment by Solxanich** [comment]
Highlights from Discord for reload button:
1) mods are auto-reloaded when leaving the menu if there are changes already. 
2) The button is thus a Force Reload button
3) Does anyone know why a player would need to Force Reload? No, we guess only modders have use of this, so hide it behind Developer mode

**Comment by jackbondpreston** [comment]
I'll take a look at trying the redesign/check out your PR later today as I agree it would be better.

In terms of removing the button, I don't mind either way but I'll let people more knowledgeable than me make that call - feel free to give any opinions on that

**Comment by Steviegt6** [comment]
> I'll take a look at trying the redesign/check out your PR later today as I agree it would be better.
> 
> In terms of removing the button, I don't mind either way but I'll let people more knowledgeable than me make that call - feel free to give any opinions on that

I think the discussion should be brought up in Discord again while this time keeping in mind that normally you can enable and disable automatic reloading, and there are actual uses in terms of saving time, hassle, and resouces.

File: patches/tModLoader/Terraria/ModLoader/ModLoader.cs [hunk 2]
```diff
@@ -195,6 +202,86 @@ private static void Load(CancellationToken token = default)
...
+						{"modloaderversion", BuildInfo.versionedName},
+						{"platform", ModLoader.CompressedPlatformRepresentation},
+						{"netversion", FrameworkVersion.Version.ToString()},
+						{"EarlyAutoUpdate", UI.ModBrowser.UIModBrowser.EarlyAutoUpdate.ToString()}
+					};
+					using (var client = new WebClient()) {
+						ServicePointManager.ServerCertificateValidationCallback = (sender, certificate, chain, policyErrors) => { return true; };
+						client.UploadValuesCompleted += UpdateCheckComplete;
+						client.UploadValuesAsync(new Uri(url), "POST", values);
+					}
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/ModLoader.cs:227]:
```suggestion
					ServicePointManager.Expect100Continue = false;
					
					const string Url = "http://javid.ddns.net/tModLoader/listmods.php";
					
					var values = new NameValueCollection {
						{ "modloaderversion", BuildInfo.versionedName },
						{ "platform", ModLoader.CompressedPlatformRepresentation },
						{ "netversion", FrameworkVersion.Version.ToString() },
						{ "EarlyAutoUpdate", UI.ModBrowser.UIModBrowser.EarlyAutoUpdate.ToString() }
					};
					
					using var client = new WebClient();
					
					ServicePointManager.ServerCertificateValidationCallback = (sender, certificate, chain, policyErrors) => true;
					client.UploadValuesCompleted += UpdateCheckComplete;
					
					client.UploadValuesAsync(new Uri(Url), "POST", values);
```

**Review by Mirsario — CHANGES_REQUESTED** [review]
I think it would be better to display the available mod updates to the bottom left corner of the screen instead of having them on the mod browser button (which doesn't even exist anymore? Merge upstream.)

Show `X Mods Enabled (Y Updates Available!)` below TML's version info and patreon link.

**Comment by jackbondpreston** [comment]
> 2\. Not everyone wants to immediately reload their mods, especially when they have larger mods enabled. I find myself often wanting to not immediately reload if I have to update mods or have others I wish to install (entirely separate from when I develop mods).
>     I don't see any reason to get rid of a perfectly-functioning feature.

How would you feel about a setting which enables auto reloading mods which require a reload upon exiting the Mods screen? When disabled the Force Reload button could be shown instead.

**Comment by jackbondpreston** [comment]
Ok, I have pushed the functionality I described above:
New config option `AutomaticallyReloadRequiredModsLeavingModsScreen` which enables/disables automatically reloading mods that require a reload when the user exits the mod screen. If this option is disabled, the Force Reload button will be displayed. If it is enabled, the button is hidden.

**Comment by jackbondpreston** [comment]
Hopefully I addressed your suggestions @Mirsario.

I guess it would be best if this change only checks for mod updates and displays the mod updates available text when a GOG instance is running (the x mods loaded could stay for Steam)?

**Pull request merged**